### ISSUE TITLE
Project bootstraping protocol

### ISSUE THREAD

**Issue by rexim** [tsoding/Issuestant#19]
Consider the following use case.

- I create a bootstrap issue for a newly created project (similar to this https://github.com/morganey-lang/morganey-mode/issues/1)
- I enlist all of the things I want in the description.
- I mention @Issuestant in the bootstrap issue.
- @Issuestant recognizes the type of the project and the enlisted stuff (from existing presets probably)
- @Issuestant bootstraps the project via series of issues and PRs. For example:
  - @Issuestant setups Travis CI
  - @Issuestant files an issue "Please enable this project in Travis CI" and assigns it to the Project Leader
  - After the Project Leader resolves the issue @Issuestant submits a PR with a predefined `.travis.yml`
- @Issuestant uses a similar protocol for the rest of the enlisted items until the project is completely bootstrapped
- If something goes wrong during the protocol @Issuestant mentions that immediately in a correspoding place and suggest possible ways of resolving a problem.
- Such protocols can be configurable.

**Comment by ForNeVeR** [comment]
**Holy Moby Dick**, do you want everything *that* automated? That's cool!

**Comment by ForNeVeR** [comment]
@rexim I'd like to discuss the implementation, _expecially_ if you already have something in mind!

**Comment by rexim** [comment]
@ForNeVeR no, I don't have anything particular in mind right now. Only abstract use cases. :) Feel free to share your ideas.

**Comment by ForNeVeR** [comment]
Well, I see the project bootstrapping workflow as a set of particular tasks (set up Jenkins, create the issue, wait for owner response etc.) **very** dispersed in time. I believe there're special programmable workflow management systems such as [JBPM](https://www.jbpm.org/) and [WWF](http://stackoverflow.com/questions/104099/when-to-use-windows-workflow-foundation) designed to deal with that (and they even have a special subroutines to deal with so-called "humanoid agent responses" that may take additional resources due to these agents' bad reaction time).

No doubt, we don't need the fully-blown workflow management systems (because they're big and nasty). Maybe there're lesser systems that will handle our current tasks? I imagine some small Clojure or Scala DSL that'll allow to start tasks and await their results.

So, I think we should look for lightweight workflow-management or (business?) process-management system. If we won't be able to find one, I'm ready to produce more detailed spec on our little implementation.

**Comment by ForNeVeR** [comment]
@rexim also I can see the risk of us involving in yet another never-ending project of creating our own workflow engine. I don't know what to do :(

Probably we need more simple ideas, and I'd like to hear them.

**Comment by rexim** [comment]
@ForNeVeR I'd suggest to just khooychit the code and see how it goes.

**Comment by rexim** [comment]
I mean, once we have at least something done it will become more clear on what we need.

**Comment by ForNeVeR** [comment]
Yes that's always a viable option.

**Comment by ForNeVeR** [comment]
Well, it's time to formalize my suggestions.

Issuestant should have a DSL (doesn't matter if it's actual DSL or just a set of specific functions) that allows the user (≡ [the programmer](https://i.ytimg.com/vi/XBGaKKoxj4E/maxresdefault.jpg)) to create *plans*.

The *plan* is a set of **instructions** to achieve the user's purpose. Every step of the plan could be either:

- send a HTTP request and check its status (e.g. set up the Travis, create an issue)
- commit something to a branch of a git repository
- *await* sentient being response (e.g. post an issue and wait for repository owner approval)

Current *plan* state should be serializable, and plan executor should persist its state after execution of every *instruction* (especially when it's *awaiting* something). **Await** itself could currently be achieved easily by just polling the resource periodically (e.g. check the issue comments every 10 minutes or so).

All the possible *instructions* (send request, git commit, await response from specific source) could be for now implemented ad-hoc (i.e. I have no suggestions to abstract the step implementations somehow). Instructions should be grouped into presets (e.g. "Setup the new git repository"); execution of specific preset could be requested by the user using github issues or whatever.

At some point we'll want support of some complex *instructions* to branch the *plan* and organize loops, but for now I don't think it's absolutely necessary.

Also it'd be nice to have some graphical interface to check *plan* execution, but we could add it at the later stage.

**Comment by rexim** [comment]
@ForNeVeR nice formalization! Thank you! I've planned this issue for Morning Tsoding (see [schedule](https://calendar.google.com/calendar/embed?src=knnbf89n7lna1q3is3nacpd054%40group.calendar.google.com&ctz=Asia/Krasnoyarsk)) Will try to see how we can split this guy.

**Comment by rexim** [comment]
Issuestant stream has been moved, sorry. https://twitter.com/tsoding/status/791568920276787200

**Comment by rexim** [comment]
http://activiti.org/
